edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/TokenHelper.cs

```diff
@@ -34,5 +38,164 @@ public static bool IsTokenExpired(string hostName, string token)
         }
 
         public static TimeSpan GetTokenExpiryTimeRemaining(string hostName, string token) => GetTokenExpiry(hostName, token) - DateTime.UtcNow;
+
+        class SharedAccessSignature
+        {
+            const string SharedAccessSignatureName = "SharedAccessSignature";
+            const string AudienceFieldName = "sr";
+            const string SignatureFieldName = "sig";
+            const string KeyNameFieldName = "skn";
+            const string ExpiryFieldName = "se";
+            const string KeyValueSeparator = "=";
+            const string PairSeparator = "&";
+            static readonly DateTime EpochTime = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
+            static readonly TimeSpan MaxClockSkew = TimeSpan.FromMinutes(5);
+
+            readonly string encodedAudience;
+            readonly string expiry;
+
+            public string IotHubName { get; }
+
+            public DateTime ExpiresOn { get; private set; }
+
+            public string KeyName { get; private set; }
+
+            public string Audience { get; private set; }
+
+            public string Signature { get; private set; }
+
+            public static SharedAccessSignature Parse(string iotHubName, string rawToken)
+            {
+                if (string.IsNullOrWhiteSpace(iotHubName))
+                {
+                    throw new ArgumentNullException(nameof(iotHubName));
+                }
+
+                if (string.IsNullOrWhiteSpace(rawToken))
+                {
+                    throw new ArgumentNullException(nameof(rawToken));
+                }
+
+                IDictionary<string, string> parsedFields = ExtractFieldValues(rawToken);
+
+                if (!parsedFields.TryGetValue(SignatureFieldName, out string signature))
+                {
+                    throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Missing field: {0}", SignatureFieldName));
+                }
+
+                if (!parsedFields.TryGetValue(ExpiryFieldName, out string expiry))
+                {
+                    throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Missing field: {0}", ExpiryFieldName));
+                }
+
+                // KeyName (skn) is optional.
+                parsedFields.TryGetValue(KeyNameFieldName, out string keyName);
+
+                if (!parsedFields.TryGetValue(AudienceFieldName, out string encodedAudience))
+                {
+                    throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Missing field: {0}", AudienceFieldName));
+                }
+
+                return new SharedAccessSignature(iotHubName, EpochTime + TimeSpan.FromSeconds(double.Parse(expiry, CultureInfo.InvariantCulture)), expiry, keyName, signature, encodedAudience);
+            }
+
+            public static bool IsSharedAccessSignature(string rawSignature)
```

Comment:
Is this needed? If not, then we can remove?

Reply:
Same for all other public methods

Reply:
the one you found already has these removed